### ISSUE TITLE
Hover on a type synonym should show its definition

### ISSUE THREAD

**Issue by cdsmith** [haskell/haskell-language-server#2222]
When I hover over a type synonym, I currently see a kind signature, and the documentation.  For example:

> `SomeType :: *`
>
> *Defined at SomeFile.hs:123:456*
>
> This is some haddock documentation.

It would make sense to see the definition, as well as (or even instead of) the kind signature.

> `SomeType :: *`
> `type SomeType = Int`
>
> *Defined at SomeFile.hs:123:456*
>
> This is some haddock documentation.

**Comment by jneira** [comment]
Not sure how easy could be access the source definition where the hover is computed (i hope it would be doable).
at least call hierarchy jumps to the type synonym definition (so it is only one or two clicks away)?

**Comment by cdsmith** [comment]
Yes, it's one click away.  I filed this issue because sometimes, when working with unfamiliar code, it would _really_ help to get access to the definition of a type synonym without having to open a new window, hide the code I was reading, and then hunt for the tab to swap back!  This would be a really big favor for users in reading synonym-heavy code.

In fact, I wish there were a good answer for navigating multiple levels of synonyms, which the code I'm reading has plenty of.  Unfortunately, hovering over the type synonym doesn't work when you're already in the hover popup.

**Comment by jneira** [comment]
A code action to desugar a type synonym recursively could help too: https://github.com/haskell/haskell-language-server/issues/506#issuecomment-923640878

**Comment by michaelpj** [comment]
I think "quickly look at a definition without navigating away" is a problem for the client, and indeed, some clients (notably VSCode and Emacs `lsp-mode`) let you "peek" at a definition, which does exactly what you want.

So I don't think we should change how hovers behave to include the definition in this one case, IMO.

**Comment by michaelpj** [comment]
Feel free to reopen if you disagree!

**Comment by Aster89** [comment]
Maybe showing the `type` definition when hovering on the it is a problem for the client, but what about when one hovers on an expression, and the type of that expression is just a single-word `type` synonym? One might want to see what's the "actual" type. I give an example at https://github.com/haskell/haskell-language-server/issues/4290#issuecomment-4808408832.
